libexec/custom_facts.rb

```diff
@@ -29,7 +30,33 @@
     $LOAD_PATH << dir unless $LOAD_PATH.include?(dir)
   end
 
+  if (conn_info = args['_target'])
+
+    unless conn_info['type']
+      puts "Cannot collect facts for a remote target without knowing it's type."
+      exit 1
+    end
+    # Some valid bolt urls like file:///tmp/foo don't have a host which means we don't have a name
+    unless conn_info['host']
+      puts "Cannot collect facts for a remote target with a host in it's name"
+    end
+
+    # TODO: validate_target
+    special_keys = %w[type debug uri]
+    connection = conn_info.reject { |k, _| special_keys.include?(k) }
+    device = OpenStruct.new(connection)
+    device.url = conn_info['uri']
+    device.provider = conn_info['type']
+    device.options = { debug: true } if conn_info['debug']
+    Puppet[:facts_terminus] = :network_device
+    Puppet[:certname] = device.name
+    Puppet::Util::NetworkDevice.init(device)
```

Comment:
This doesn't work well with modules that subclass the `NetworkDevice::Simple` from resource API because it means tasks that work like the panos PR are more flexible in terms of config format while apply only works when the name is set to the devices url. 

Solutions:
1. Do nothing. if you want to use apply you have to rely on the url, all `NetworkDevice` modules should document only this format. When the Transport API is added to to resource_api we can use more flexible config for them.
2. Check for subclasses of `NetworkDevice::Simple` and bypass `NetworkDevice.init` for them.

I think there is a related question of how should we differentiate "network_device" modules from "transport" based remote modules.